src/uu/tail/src/chunks.rs

```diff
@@ -81,3 +91,630 @@ impl<'a> Iterator for ReverseChunks<'a> {
+            if lines_offset == 0 {
+                break;
+            }
+            if byte == &self.delimiter {
+                lines_offset -= 1;
+            }
+            bytes_offset += 1;
+        }
+        bytes_offset
+    }
+
+    /// Print the bytes contained in this buffer calculated with the given offset in number of
+    /// lines.
+    ///
+    /// # Arguments
+    ///
+    /// * `writer`: must implement [`Write`]
+    /// * `offset`: An offset in number of lines.
+    pub(crate) fn print_lines(&self, writer: &mut impl Write, offset: usize) -> UResult<()> {
+        self.print_bytes(writer, self.calculate_bytes_offset_from(offset))
+    }
+
+    /// Print the bytes contained in this buffer beginning from the given offset in number of bytes.
+    ///
+    /// # Arguments
+    ///
+    /// * `writer`: must implement [`Write`]
+    /// * `offset`: An offset in number of bytes.
+    pub(crate) fn print_bytes(&self, writer: &mut impl Write, offset: usize) -> UResult<()> {
+        writer.write_all(self.get_buffer_with(offset))?;
+        Ok(())
+    }
+}
+
+/// An abstraction layer on top of [`LinesChunk`] mainly to simplify filling only the needed amount
+/// of chunks. See also [`Self::fill`]. Works similar like [`BytesChunkBuffer`], but works on top
+/// of lines delimited by [`Self::delimiter`] instead of bytes.
+pub struct LinesChunkBuffer {
+    // The delimiter to recognize a line. Any [`u8`] is allowed.
+    delimiter: u8,
+    // The amount of lines occurring in all currently stored [`LinesChunk`]s
+    lines: u64,
+    // The amount of lines to print.
+    num_print: u64,
+    // Stores the [`LinesChunk`]
+    chunks: VecDeque<Box<LinesChunk>>,
+}
+
+impl LinesChunkBuffer {
+    /// Create a new [`LinesChunkBuffer`]
+    pub(crate) fn new(delimiter: u8, num_print: u64) -> Self {
+        Self {
+            delimiter,
+            num_print,
+            lines: 0,
+            chunks: VecDeque::new(),
+        }
+    }
+
+    /// Fills this buffer with chunks and consumes the reader completely. This method ensures that
+    /// there are exactly as many chunks as needed to match [`Self::num_print`] lines, so there are
+    /// in sum exactly `self.num_print` lines stored in all chunks. The method returns an iterator
+    /// over these chunks. If there are no chunks, for example because the piped stdin contained no
+    /// lines, or `num_print = 0` then `iterator.next` will return None.
+    ///
+    /// # Arguments
+    ///
+    /// * `reader`: A [`BufReader`] with an inner element implementing the [`Read`] trait.
+    ///
+    /// returns: Result<Iter<Box<LinesChunk, Global>>, Box<dyn UError, Global>>
+    pub(crate) fn fill(
+        &mut self,
+        reader: &mut BufReader<impl Read>,
+    ) -> UResult<Iter<Box<LinesChunk>>> {
+        let mut chunk = Box::new(LinesChunk::new(self.delimiter));
+
+        while (chunk.fill(reader)?).is_some() {
+            self.lines += chunk.lines as u64;
```

Comment:
Maybe `self.lines` could be `usize`? It seems to be used with other `usize` values a lot.

Reply:
you're right, the casting looks cumbersome but I wanted to support tailing files > 4GB on 32-bit systems, therefore the difference of LinesChunkBuffer.lines as u64 and LinesChunk.lines as usize. The u64 originates from FilterMode::Lines(count, ...) and FilterMode::Bytes(count), both `count` variables being u64. The usize is sufficient for LinesChunk.lines because the max value for lines is bytes, and bytes max value is BUFFER_SIZE which is a usize. Would it help updating the docs to include some infos about the choice of LinesChunkBuffer.lines as u64?